flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java

```diff
@@ -0,0 +1,242 @@
+
+	private final Logger log = LoggerFactory.getLogger(DispatcherJob.class);
+
+	private final CompletableFuture<JobManagerRunner> jobManagerRunnerFuture;
+	private final CompletableFuture<ArchivedExecutionGraph> jobResultFuture;
+
+	private final long initializationTimestamp;
+	private final JobID jobId;
+	private final String jobName;
+
+	private final Object lock = new Object();
+
+	// internal field to track job status during initialization. Is not updated anymore after
+	// job is initialized, cancelled or failed.
+	@GuardedBy("lock")
+	private JobStatus jobStatus = JobStatus.INITIALIZING;
+
+	private final CompletableFuture<Void> terminationFuture = new CompletableFuture<>();
+
+	private enum SubmissionType {
+		INITIAL, RECOVERY
+	}
+
+	static DispatcherJob createForSubmission(
+		CompletableFuture<JobManagerRunner> jobManagerRunnerFuture,
+		JobID jobId,
+		String jobName,
+		long initializationTimestamp) {
+		return new DispatcherJob(jobManagerRunnerFuture, jobId, jobName, initializationTimestamp, SubmissionType.INITIAL);
+	}
+
+	static DispatcherJob createForRecovery(
+		CompletableFuture<JobManagerRunner> jobManagerRunnerFuture,
+		JobID jobId,
+		String jobName,
+		long initializationTimestamp) {
+		return new DispatcherJob(jobManagerRunnerFuture, jobId, jobName, initializationTimestamp, SubmissionType.RECOVERY);
+	}
+
+	private DispatcherJob(
+		CompletableFuture<JobManagerRunner> jobManagerRunnerFuture,
+		JobID jobId,
+		String jobName,
+		long initializationTimestamp,
+		SubmissionType submissionType) {
+		this.jobManagerRunnerFuture = jobManagerRunnerFuture;
+		this.jobId = jobId;
+		this.jobName = jobName;
+		this.initializationTimestamp = initializationTimestamp;
+		this.jobResultFuture = new CompletableFuture<>();
+
+		FutureUtils.assertNoException(this.jobManagerRunnerFuture.handle((jobManagerRunner, throwable) -> {
+			// JM has been initialized, or the initialization failed
+			synchronized (lock) {
+				if (throwable == null) {
+					// Forward result future
+					FutureUtils.forward(jobManagerRunner.getResultFuture(), jobResultFuture);
+
+					if (jobStatus == JobStatus.CANCELLING) {
+						log.info("Cancellation during initialization has been requested for job {}. Initialization completed, cancelling job.", jobId);
+
+						// cancel job
+						jobManagerRunner
+							.getJobMasterGateway()
+							.thenCompose(gw -> gw.cancel(RpcUtils.INF_TIMEOUT))
+							.whenComplete((ignored, cancelThrowable) -> {
+							if (cancelThrowable != null) {
+								log.warn("Cancellation of job {} failed", jobId, cancelThrowable);
+							}
+						});
+
+						// cancellation will eventually complete the jobResultFuture
+						jobResultFuture.whenComplete((archivedExecutionGraph, resultThrowable) -> {
+							synchronized (lock) {
+								if (resultThrowable == null) {
+									jobStatus = archivedExecutionGraph.getState();
+								} else {
+									jobStatus = JobStatus.FAILED;
+								}
+							}
+						});
```

Comment:
> What happens if I call dispatcherJob.requestJob after the job has been cancelled in state INITIALIZING? I would like to obtain the final CANCELED ArchivedExecutionGraph ideally.

Since we are forwarding the the result future from the runner to the `DispatcherJob`'s result future, you should get a proper `ArchivedExecutionGraph` with the correct final state.

(note: I have removed this entire `.whenComplete()` stage, as it is not necessary to track the FAILED or CANCELLED etc. states).